### ISSUE TITLE
RFC: saying sorry

### ISSUE THREAD

**Issue by gebner** [leanprover/lean3#1355]
Once upon a time, `sorry` was only used if the user actually typed `sorry` in a Lean file.  However, more and more failure conditions are now being excused by `sorry` as well--usually in a laudable and worthwhile effort to recover from parsing and elaboration errors.
 1. As intended, we use `sorry` when the user types `sorry`.
 2. We also use `sorry` as a placeholder for the value of a definition when elaboration fails.
 3. We might use `sorry` in the future as a placeholder in expressions that have parse errors (e.g. `(if x > 0 then tt)` would be parsed as `if x > 0 then tt else sorry`)
 4. While developing proof automation, we might want to insert `sorry` as proof terms.

Aside from 1., all these applications have one issue in common: we cannot syntactically determine from the pre-expression whether we will use sorry or not.  Up until now, we added a `sorry` axiom to the environment the first time it was used.  This worked well when Lean was single-threaded and the main source for `sorry`s was explicit use by the user.  However now `sorry` may first appear out of nowhere in a lemma that is elaborated asynchronously, and at that point we can no longer add it to the environment.
Hence I believe that we should just proactively add the sorry axiom as the first declaration to the environment.  I propose that we don't add the sorry axiom in the standard library but declare it in the C++ code instead, then it doesn't get exported and we can easily disable it using an option.

Another issue is that how we handle the computability lint for sorry: because sorry is noncomputable, we would normally get errors in all dependent definitions (because we would usually have to mark them noncomputable as well).  As a workaround, if the parser sees sorry (if the user enters it) we set the "noncomputable theory" flag for the current file.  Other files still show the noncomputability errors.  Actual noncomputability errors are hidden.
I think it would be much simpler if we just had a special case for sorry, and made it computable by fiat.  (We already have VM code for it!)

Concretely, I propose the following changes:
1. Add a global option `allow_sorry`.  This option is set to true by default, at least interactively.
2. If `allow_sorry` is set to to true, we add a `sorry` axiom to the initial environment.
3. The `noncomputable` extension treats `sorry` as computable.
4. We report a warning when we add a declaration to the environment that uses sorry.
5. The type checkers do not add the sorry axiom.

The main objection I can see is that `print axioms` would now always print one more axiom (by default), namely sorry.  But if I understand the plan for the native code interface correctly, we will get new constants and axioms anyhow--for example the constant `io` and axioms describing the properties of the IO monad.

**Comment by leodemoura** [comment]
What about a `sorry` macro?
One nice side-effect is that trust level 0 will reject files that use `sorry`.
So, if the standard library has a temporary sorry, the new test you have added will break. Depending on how we see this it may be a good/bad thing.

**Comment by leodemoura** [comment]
> Another issue is that how we handle the computability lint for sorry: because sorry is noncomputable, we would normally get errors in all dependent definitions (because we would usually have to mark them noncomputable as well). As a workaround, if the parser sees sorry (if the user enters it) we set the "noncomputable theory" flag for the current file. 

Another option: we skip the test if the definition depends on `sorry`.

I'm not claiming it is a better solution. I'm just listing possibilities.

**Comment by gebner** [comment]
I like the sorry macro.  It would be computable as well.

> So, if the standard library has a temporary sorry, the new test you have added will break. Depending on how we see this it may be a good/bad thing.

The test will already fail at the export stage, since the export format doesn't support sorry.
I believe that the type checkers shouldn't accept sorry anyhow.  Their purpose is to make sure a theorem is true--and sorry doesn't help with that.

**Comment by leodemoura** [comment]
Your proposal has an advantage. `print axioms foo` will include `sorry` if `foo` depends on `sorry`.
I believe this is why we did not use a `sorry` macro in the current implementation.

BTW, in both options, the module importer will have to check whether imported declarations use `sorry` because we want to report a warning when we import something that uses `sorry`. In the current implementation, this is trivial since we just check whether the environment contains `sorry`. 
We may save this information in the .olean file to be able to skip the test.

**Comment by leodemoura** [comment]
> The main objection I can see is that print axioms would now always print one more axiom (by default), namely sorry.

I missed this one. You think it is bad to include `sorry`.

**Comment by gebner** [comment]
> Your proposal has an advantage. print axioms foo will include sorry if foo depends on sorry.

This should be a one-liner in print_axioms_deps.

> BTW, in both options, the module importer will have to check whether imported declarations use sorry because we want to report a warning when we import something that uses sorry. 

We could add an environment extension that keeps track of it.  Then this information would be easily available in the environment and trivially serialized in the olean file.
We can add this check in `module::add`, but then it can be easily circumvented by calling `environment::add` directly.

**Comment by gebner** [comment]
> I missed this one. You think it is bad to include sorry.

Only in the default `print axioms` command, without any names afterwards.  Right now, it returns a list of all axioms in the environment--with my first proposal it would look like this:
```
quot.sound : ∀ {α : Type u} {r : α → α → Prop} {a b : α}, r a b → quot.mk r a = quot.mk r b
classical.strong_indefinite_description : Π {a : Sort u} (p : a → Prop), nonempty a → {x // (∃ (y : a), p y) → p x}
sorry : Π {A : Sort l}, A
propext : ∀ {a b : Prop}, (a ↔ b) → a = b
```
But I'm not too concerned, since we will soon have `io`, `io.bind`, `io.is_monad`, etc., there as well.

**Comment by leodemoura** [comment]
> But if I understand the plan for the native code interface correctly, we will get new constants and axioms anyhow--for example the constant io and axioms describing the properties of the IO monad.

I'm going off topic. I had lunch with Nik (@nikswamy) this week and he suggested a clean way of supporting IO and ST monads in Lean without adding a bunch of axioms. The idea is to use a structure that abstracts their APIs. Example:
```lean
class io_class :=
(m : Type -> Type)
(put_str : string -> m unit)
(get_line : m string)
...
def io [io_class] : Type -> Type := io_class.m
```
and only the top level is capable of providing the `io_class` magical instance.

**Comment by gebner** [comment]
The `io_class` is nice.  You could even define a mock one for testing.  I just wonder how extensible it is in practice.  I guess we need one field per "syscall", or can we extend it in another way?

**Comment by leodemoura** [comment]
> I just wonder how extensible it is in practice. I guess we need one field per "syscall", or can we extend it in another way?

Yes, it is not very extensible. I think we can try to define small classes for related `syscall`s, and then group all of them into `io_class`.

```lean
class file_system (m : Type -> Type) :=
...

class process (m : Type -> Type) :=
...

class io_class :=
(m  : Type -> Type)
(fs  : file_system m)
(pr  : process m)
...
```

We also have to do this grouping for performance reasons because a structure with `n` fields takes `O(n^2)` to elaborate. The recursor `rec` has size `O(n)`, and each automatically generated projection uses `rec` and has size `O(n)`.

**Comment by gebner** [comment]
This issue has been solved by making sorry a macro in PR #1358.
